File: backend/app/models/interview.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import CheckConstraint, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import GUID, JSONType, BaseModel

if TYPE_CHECKING:
    from app.models.role import Role, Skill
    from app.models.user import User


class InterviewSession(BaseModel):
    __tablename__ = "interview_sessions"
    __table_args__ = (
        CheckConstraint(
            "status IN ('configured', 'in_progress', 'completed', 'abandoned')",
            name="ck_session_status",
        ),
        CheckConstraint(
            "experience_level IN ('fresher', 'junior', 'mid', 'senior')",
            name="ck_experience_level",
        ),
        CheckConstraint(
            "difficulty IN ('easy', 'medium', 'hard', 'adaptive')",
            name="ck_difficulty",
        ),
    )

    user_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("users.id", ondelete="CASCADE"), nullable=False
    )
    role_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("roles.id"), nullable=False
    )
    experience_level: Mapped[str] = mapped_column(String(20), nullable=False)
    difficulty: Mapped[str] = mapped_column(String(20), nullable=False, default="adaptive")
    duration_minutes: Mapped[int] = mapped_column(Integer, nullable=False, default=30)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="configured")
    focus_areas: Mapped[dict | None] = mapped_column(JSONType(), nullable=True)
    question_budget: Mapped[int] = mapped_column(Integer, nullable=False, default=10)
    questions_asked: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    user: Mapped[User] = relationship(back_populates="interview_sessions", lazy="joined")
    role: Mapped[Role] = relationship(lazy="joined")
    questions: Mapped[list[InterviewQuestion]] = relationship(
        back_populates="session", lazy="noload", cascade="all, delete-orphan"
    )


class InterviewQuestion(BaseModel):
    __tablename__ = "interview_questions"

    session_id: Mapped[uuid.UUID] = mapped_column(
        GUID(),
        ForeignKey("interview_sessions.id", ondelete="CASCADE"),
        nullable=False,
    )
    skill_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(), ForeignKey("skills.id"), nullable=True
    )
    sequence_number: Mapped[int] = mapped_column(Integer, nullable=False)
    question_text: Mapped[str] = mapped_column(Text, nullable=False)
    difficulty: Mapped[str] = mapped_column(String(20), nullable=False, default="medium")
    question_type: Mapped[str] = mapped_column(String(20), nullable=False, default="initial")
    parent_question_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(),
        ForeignKey("interview_questions.id"),
        nullable=True,
    )
    metadata_: Mapped[dict | None] = mapped_column("metadata", JSONType(), nullable=True)

    session: Mapped[InterviewSession] = relationship(back_populates="questions", lazy="joined")
    skill: Mapped[Skill | None] = relationship(lazy="joined")
    answer: Mapped[InterviewAnswer | None] = relationship(
        back_populates="question", lazy="selectin", uselist=False
    )
    parent_question: Mapped[InterviewQuestion | None] = relationship(
        remote_side="InterviewQuestion.id", lazy="noload"
    )


class InterviewAnswer(BaseModel):
    __tablename__ = "interview_answers"

    question_id: Mapped[uuid.UUID] = mapped_column(
        GUID(),
        ForeignKey("interview_questions.id", ondelete="CASCADE"),
        unique=True,
        nullable=False,
    )
    answer_text: Mapped[str] = mapped_column(Text, nullable=False)
    response_time_seconds: Mapped[int | None] = mapped_column(Integer, nullable=True)

    question: Mapped[InterviewQuestion] = relationship(back_populates="answer", lazy="joined")

File: frontend/src/types/interview.ts
export interface InterviewConfig {
  user_id?: string;
  role_id: string;
  experience_level: ExperienceLevel;
  difficulty: Difficulty;
  duration_minutes: number;
  focus_areas?: string[];
}

export type ExperienceLevel = 'fresher' | 'junior' | 'mid' | 'senior';
export type Difficulty = 'easy' | 'medium' | 'hard' | 'adaptive';

export interface InterviewSession {
  id: string;
  user_id: string;
  role: { id: string; name: string };
  experience_level: string;
  difficulty: string;
  duration_minutes: number;
  question_budget: number;
  focus_areas: string[] | null;
  status: string;
  created_at: string;
}

export interface Question {
  id: string;
  sequence_number: number;
  question_text: string;
  difficulty: string;
  skill: string;
  question_type: string;
  parent_question_id?: string | null;
}

export interface Progress {
  current: number;
  total: number;
  skills_covered: string[];
  skills_remaining: string[];
}

export interface InterviewStartResponse {
  session_id: string;
  status: string;
  interviewer_message: string;
  question: Question;
  progress: Progress;
}

export interface EvaluationResponse {
  question_id: string;
  overall_score: number;
  technical_correctness: number;
  conceptual_depth: number;
  communication_clarity: number;
  relevance: number;
  problem_solving: number;
  completeness: number;
  feedback: string;
  strengths: string[];
  weaknesses: string[];
}

export interface AnswerRequest {
  question_id: string;
  answer_text: string;
  response_time_seconds?: number;
}

export interface AnswerResponse {
  evaluation: EvaluationResponse;
  next_question: Question | null;
  progress: Progress;
  interview_complete: boolean;
  closing_message: string | null;
}

export interface InterviewCompleteResponse {
  session_id: string;
  status: string;
  questions_asked: number;
  question_budget: number;
  message: string;
}

export interface ConversationEntry {
  type: 'question' | 'answer' | 'system';
  text: string;
  question?: Question;
  evaluation?: EvaluationResponse;
  timestamp: number;
}

export interface QuestionEvaluation {
  overall_score: number;
  feedback: string;
  strengths: string[];
  weaknesses: string[];
}

export interface ReportQuestion {
  sequence_number: number;
  question_text: string;
  answer_text: string;
  evaluation: QuestionEvaluation;
}

export interface ReportResponse {
  id: string;
  session_id: string;
  overall_score: number;
  technical_score: number;
  communication_score: number;
  problem_solving_score: number;
  confidence_score: number;
  readiness_level: string;
  summary: string;
  strengths: string[];
  weaknesses: string[];
  recommendations: string[];
  recommended_topics: string[];
  category_breakdown: Record<string, number>;
  questions: ReportQuestion[];
  created_at: string;
}

export interface InterviewHistoryItem {
  id: string;
  role_name: string;
  experience_level: string;
  status: string;
  overall_score: number | null;
  questions_asked: number;
  started_at: string | null;
  completed_at: string | null;
}

export interface InterviewHistoryResponse {
  interviews: InterviewHistoryItem[];
  total: number;
  limit: number;
  offset: number;
}
